SilverSolver/assignment2/binary_tree.py

```diff
@@ -0,0 +1,61 @@
+class BinaryNode:
+
+    def __init__(self, value, parent=None, child1=None, child2=None):
+        self.value  = value
+        self.parent = parent
+        self.child1 = child1
+        self.child2 = child2
+
+    def get_ancestors(self, prev_ancestors=None, value=True):
+        if not prev_ancestors:
+            prev_ancestors = []
+        if not self.parent:
+            return prev_ancestors
+        else:
+            if value:
+                prev_ancestors.append(self.parent.value)
+            else:
+                prev_ancestors.append(self.parent)
+            return self.parent.get_ancestors(prev_ancestors=prev_ancestors, value=value)
+
+    def get_lowest_common_ancestor(self, another_node, value=True):
+        self_anc = self.get_ancestors(value=False)
+        another_anc = another_node.get_ancestors(value=False)
+        low_c_anc = None
+        while(self_anc and another_anc):
+            if self_anc[-1] == another_anc[-1]:
+                low_c_anc = self_anc[-1]
+                self_anc = self_anc[:-1]
+                another_anc = another_anc[:-1]
+            else:
+                if value and low_c_anc:
```

Comment:
I suggest you go with the value=True everywhere.

Reply:
I suggest that two **different** nodes can store the same values, but always have different addresses in RAM/pointers etc.
I motivate this approach by the fact that nodes with the same value can have different parents and children, so they can have different positions in the tree; but nodes with the same address in RAM are factually the same object with the same position and all and should be considered as the same node.
If I understood the assignment correctly, that in the **first** part I should return the **values** of ancestors of the node. But in the **second** part I should return the value of **common ancestor** . So I thought that I should find the _object_, which is the common ancestor of two nodes in the tree structure. That's why I thought that I should have some option to store and comparing ancestor _objects_ itself (i.e. pointers to objects), not only their values to see the difference between possibly different ancestors with the same value. And that's why I added the possibility to get pointers to ancestors (if value=False), not only their values (which is going on if value=True).

Let's consider a little example.
0
| \
1 1
|  |
2 2
If I will use all my code with value=True at this example, it will say that the lowest common ancestor for least nodes (with values 2 and 2) is "1", which have no sense in my opinion, as I tried to explain before. But if I will use all my code as I use it now (i.e. with value=False when I need this), it will say that the lowest common ancestor is "0", which seems reasonable.

Sorry if this explanation seems too excess to you. My English isn't very good, that's why I tried express an idea in as much detail as possible, to neutralize possible misunderstanding. I hope that it is clear now!

Reply:
Oh, I see now at your variant of solution that you made assumption that values held by the nodes are unique! That's where point of our divergence is. :smile: